File: multi-threaded-santa/python/src/multi_threaded_santa/present.py
from enum import Enum, auto


class PresentState(Enum):
    UNMADE = auto()
    MADE = auto()
    WRAPPED = auto()
    LOADED = auto()
    DELIVERED = auto()


class Present:
    def __init__(self, present_id: int) -> None:
        self.id = present_id
        self._state = PresentState.UNMADE

    @property
    def state(self) -> PresentState:
        return self._state

    def make(self) -> None:
        if self._state != PresentState.UNMADE:
            raise RuntimeError(f"Cannot make a present in state {self._state.name}.")
        self._state = PresentState.MADE

    def wrap(self) -> None:
        if self._state != PresentState.MADE:
            raise RuntimeError(f"Cannot wrap a present in state {self._state.name}.")
        self._state = PresentState.WRAPPED

    def load(self) -> None:
        if self._state != PresentState.WRAPPED:
            raise RuntimeError(f"Cannot load a present in state {self._state.name}.")
        self._state = PresentState.LOADED

    def deliver(self) -> None:
        if self._state != PresentState.LOADED:
            raise RuntimeError(f"Cannot deliver a present in state {self._state.name}.")
        self._state = PresentState.DELIVERED

File: multi-threaded-santa/csharp/src/MultiThreadedSanta/Present.cs
namespace MultiThreadedSanta;

public class Present
{
    public int Id { get; }
    public PresentState State { get; private set; } = PresentState.Unmade;

    public Present(int id)
    {
        Id = id;
    }

    public void Make()
    {
        if (State != PresentState.Unmade)
            throw new InvalidOperationException($"Cannot make a present in state {State}.");
        State = PresentState.Made;
    }

    public void Wrap()
    {
        if (State != PresentState.Made)
            throw new InvalidOperationException($"Cannot wrap a present in state {State}.");
        State = PresentState.Wrapped;
    }

    public void Load()
    {
        if (State != PresentState.Wrapped)
            throw new InvalidOperationException($"Cannot load a present in state {State}.");
        State = PresentState.Loaded;
    }

    public void Deliver()
    {
        if (State != PresentState.Loaded)
            throw new InvalidOperationException($"Cannot deliver a present in state {State}.");
        State = PresentState.Delivered;
    }
}
